shadowsocks/stream.go

```diff
@@ -187,49 +187,51 @@ func (sr *shadowsocksReader) readBlock(decryptedBlockSize int) ([]byte, error) {
 }
 
 func (sr *shadowsocksReader) Read(b []byte) (int, error) {
-	n, err := sr.readLoop(b)
-	return int(n), err
+	if err := sr.populateLeftover(); err != nil {
+		return 0, err
+	}
+	n := copy(b, sr.leftover)
+	sr.leftover = sr.leftover[n:]
+	return n, nil
 }
 
 func (sr *shadowsocksReader) WriteTo(w io.Writer) (written int64, err error) {
-	n, err := sr.readLoop(w)
-	if err == io.EOF {
-		err = nil
+	for {
+		if err = sr.populateLeftover(); err != nil {
+			if err == io.EOF {
+				err = nil
+			}
+			return written, err
+		}
+		n, err := w.Write(sr.leftover)
+		written += int64(n)
+		sr.leftover = sr.leftover[n:]
+		if err != nil {
+			return written, err
+		}
 	}
-	return n, err
 }
 
-func (sr *shadowsocksReader) readLoop(w interface{}) (written int64, err error) {
-	for {
-		if len(sr.leftover) == 0 {
-			buf, err := sr.readBlock(2)
-			if err != nil {
-				if err != io.EOF && err != io.ErrUnexpectedEOF {
-					err = fmt.Errorf("failed to read payload size: %v", err)
-				}
-				return written, err
-			}
-			size := (int(buf[0])<<8 + int(buf[1])) & payloadSizeMask
-			payload, err := sr.readBlock(size)
-			if err != nil {
-				return written, fmt.Errorf("failed to read payload: %v", err)
+// Ensures that sr.leftover is nonempty.  If leftover is empty, this method
+// waits for incoming data and decrypts it.
+// Returns an error only if sr.leftover could not be populated.
+func (sr *shadowsocksReader) populateLeftover() error {
```

Comment:
Maybe call this `ensureLeftover()`, since it doesn't always populate it? It will align with your comment ("Ensures....")

Reply:
Done